src/modules/rust.rs

```diff
@@ -19,27 +23,85 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
         return None;
     }
 
-    match get_rust_version() {
-        Some(rust_version) => {
-            const RUST_CHAR: &str = "🦀 ";
+    let module_version = if let Some(toolchain) = env_rustup_toolchain()
+        .or_else(|| execute_rustup_override_list(&context.current_dir))
+        .or_else(|| find_rust_toolchain_file(&context.current_dir))
+    {
+        match execute_rustup_run_rustc_version(&toolchain) {
+            Err(err) => err,
+            Ok(stdout) => format_rustc_version(stdout),
+        }
+    } else {
+        format_rustc_version(execute_rustc_version()?)
+    };
+
+    let mut module = context.new_module("rust");
 
-            let mut module = context.new_module("rust");
-            let module_style = module
-                .config_value_style("style")
-                .unwrap_or_else(|| Color::Red.bold());
-            module.set_style(module_style);
+    let module_style = module
+        .config_value_style("style")
+        .unwrap_or_else(|| Color::Red.bold());
+    module.set_style(module_style);
+    module.new_segment("symbol", RUST_CHAR);
+    module.new_segment("version", &module_version);
 
-            let formatted_version = format_rustc_version(rust_version);
-            module.new_segment("symbol", RUST_CHAR);
-            module.new_segment("version", &formatted_version);
+    Some(module)
+}
 
-            Some(module)
+fn env_rustup_toolchain() -> Option<String> {
+    let val = env::var("RUSTUP_TOOLCHAIN").ok()?;
+    Some(val.trim().to_owned())
+}
+
+fn execute_rustup_override_list(cwd: &Path) -> Option<String> {
+    let Output { stdout, .. } = Command::new("rustup")
+        .args(&["override", "list"])
+        .output()
+        .ok()?;
+    let stdout = String::from_utf8(stdout).ok()?;
+    if stdout == "no overrides\n" {
+        return None;
+    }
+    stdout
+        .lines()
+        .flat_map(|line| {
+            let mut words = line.split_whitespace();
+            let dir = words.next()?;
+            let toolchain = words.next()?;
+            Some((dir, toolchain))
+        })
+        .find(|(dir, _)| cwd.starts_with(dir))
+        .map(|(_, toolchain)| toolchain.to_owned())
+}
+
+fn find_rust_toolchain_file(mut dir: &Path) -> Option<String> {
+    loop {
+        let path = dir.join("rust-toolchain");
+        if path.exists() {
+            let content = fs::read_to_string(path).ok()?;
+            break Some(content.trim().to_owned());
         }
-        None => None,
+        dir = dir.parent()?;
+    }
+}
+
+fn execute_rustup_run_rustc_version(toolchain: &str) -> Result<String, String> {
+    let output = Command::new("rustup")
+        .args(&["run", toolchain, "rustc", "--version"])
+        .output()
+        .map_err(|e| e.to_string())?;
+    if output.status.success() {
+        String::from_utf8(output.stdout).map_err(|e| e.to_string())
+    } else {
+        let stderr = String::from_utf8(output.stderr).map_err(|e| e.to_string())?;
```

Comment:
I'm personally not a huge fan of the long error in the prompt. Can we throw in some config to toggle this? Perhaps we could hide the whole module if this fails, or perhaps a single "alert" symbol to indicate something is wrong but without saying what.

Reply:
Now it displays toolchain name (`channel-triple`). I updated the screenshot.

Reply:
Great!